x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/dataframe/DataFrameAnalyticsConfig.java

```diff
@@ -318,8 +342,37 @@ public Builder setHeaders(Map<String, String> headers) {
             return this;
         }
 
+        public Builder setModelMemoryLimit(ByteSizeValue modelMemoryLimit) {
+            if (modelMemoryLimit != null && modelMemoryLimit.compareTo(MIN_MODEL_MEMORY_LIMIT) < 0) {
+                throw new IllegalArgumentException("[" + MODEL_MEMORY_LIMIT.getPreferredName()
+                    + "] must be at least [" + MIN_MODEL_MEMORY_LIMIT.getStringRep() + "].");
+            }
+            this.modelMemoryLimit = modelMemoryLimit;
+            return this;
+        }
+
+        public DataFrameAnalyticsConfig applyMaxModelMemoryLimitAndBuild(ByteSizeValue maxModelMemoryLimit) {
```

Comment:
Thinking about how this works, I have a feeling it might be nice to try the idea of passing the max limit in the builder constructor. It can be an optional constructor as it is only meant to be used by the PUT action. The usage of the builder would then feel more natural.

Reply:
Good idea.  I think that will make things clearer.

Reply:
This will allow capping the default as well.